app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,555 @@
+
+      builder.setPositiveButton(this.getString(R.string.yes),
+          (dialog, id) -> openMobileDataActivity());
+      builder.setNegativeButton((android.R.string.no),
+          (dialog, id) -> startHotspotManuallyDialog());
+      builder.setTitle(this.getString(R.string.mobile_data_enabled));
+      builder.setMessage(
+          this.getString(R.string.mobile_data_message) + "\n" + this.getString(
+              R.string.mobile_data_message_confirmation)
+      );
+      builder.setCancelable(false);
+      AlertDialog dialog = builder.create();
+      dialog.show();
+    }
+  }
+
+  @Override public void onServerStarted(@NonNull String ipAddress) {
+    this.ip = ipAddress;
+    layoutServerStarted();
+  }
+
+  @Override public void onServerStopped() {
+    layoutServerStopped();
+    if (selectedBooksPath.size() > 0) {
+      selectedBooksPath.clear();
+    }
+  }
+
+  @Override public void onServerFailedToStart() {
+    Toast.makeText(this, R.string.server_failed_toast_message, Toast.LENGTH_LONG).show();
+  }
+
+  @Override public void onHotspotTurnedOn(@NonNull WifiConfiguration wifiConfiguration) {
+
+    hotspotService.startForegroundNotificationHelper();
+
+    //Show an alert dialog for hotspot details
+    AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+    builder.setPositiveButton(android.R.string.ok, (dialog, id) -> {
+
+      progressDialog =
+          ProgressDialog.show(this, getString(R.string.progress_dialog_starting_server), "",
+              true);
+      progressDialog.show();
+
+      startFlowable();
+
+    });
+
+    builder.setTitle(this.getString(R.string.hotspot_turned_on));
+    builder.setMessage(
+        this.getString(R.string.hotspot_details_message) + "\n" + this.getString(
+            R.string.hotspot_ssid_label) + " " + wifiConfiguration.SSID + "\n" + this.getString(
+            R.string.hotspot_pass_label) + " " + wifiConfiguration.preSharedKey);
+
+    builder.setCancelable(false);
+    AlertDialog dialog = builder.create();
+    dialog.show();
+
+    //setupServer();
+  }
+
+  void launchTetheringSettingsScreen() {
+    final Intent intent = new Intent(Intent.ACTION_MAIN, null);
+    intent.addCategory(Intent.CATEGORY_LAUNCHER);
+    final ComponentName cn =
+        new ComponentName("com.android.settings", "com.android.settings.TetherSettings");
+    intent.setComponent(cn);
+    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+    startActivity(intent);
+  }
+
+  @Override public void onHotspotFailedToStart() {
+    //Show a dialog to turn off default hotspot
+
+    alertDialogShower.show(KiwixDialog.TurnOffHotspotManually.INSTANCE,
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            launchTetheringSettingsScreen();
+            return Unit.INSTANCE;
+          }
+        });
+  }
+
+  @Override public void onHotspotStateReceived(@NonNull Boolean isHotspotEnabled) {
```

Comment:
the flow of this is wrong. 
The HotpotService own WifiManager, information should flow from the WifiManager to HotspotService and if necessary from HotspotService to ZimHostActivity. Instead right now it goes Activity->Service->Manager->Activity->Service

Reply:
Fixed. aee5c81baac0601b1c7da7fd925fb45ae7255d28